numba/analysis.py

```diff
@@ -0,0 +1,248 @@
+                                set())
+    escaping_dead_vars = reduce(operator.or_, escaping_dead_map.values(),
+                                set())
+    exit_dead_vars = reduce(operator.or_, exit_dead_map.values(), set())
+    dead_vars = (internal_dead_vars | escaping_dead_vars | exit_dead_vars)
+    missing_vars = all_vars - dead_vars
+    if missing_vars:
+        # There are no exit points
+        if not cfg.exit_points():
+            # We won't be able to verify this
+            pass
+        else:
+            msg = 'liveness info missing for vars: {0}'.format(missing_vars)
+            raise RuntimeError(msg)
+
+    combined = dict((k, internal_dead_map[k] | escaping_dead_map[k])
+                    for k in blocks)
+
+    return _dead_maps_result(internal=internal_dead_map,
+                             escaping=escaping_dead_map,
+                             combined=combined)
+
+
+def compute_live_variables(cfg, blocks, var_def_map, var_dead_map):
+    """
+    Compute the live variables at the beginning of each block
+    and at each yield point.
+    The ``var_def_map`` and ``var_dead_map`` indicates the variable defined
+    and deleted at each block, respectively.
+    """
+    # live var at the entry per block
+    block_entry_vars = defaultdict(set)
+
+    def fix_point_progress():
+        return tuple(map(len, block_entry_vars.values()))
+
+    old_point = None
+    new_point = fix_point_progress()
+
+    # Propagate defined variables and still live the successors.
+    # (note the entry block automatically gets an empty set)
+
+    # Note: This is finding the actual available variables at the entry
+    #       of each block. The algorithm in compute_live_map() is finding
+    #       the variable that must be available at the entry of each block.
+    #       This is top-down in the dataflow.  The other one is bottom-up.
+    while old_point != new_point:
+        # We iterate until the result stabilizes.  This is necessary
+        # because of loops in the graphself.
+        for offset in blocks:
+            # vars available + variable defined
+            avail = block_entry_vars[offset] | var_def_map[offset]
+            # substract variables deleted
+            avail -= var_dead_map[offset]
+            # add ``avail`` to each successors
+            for succ, _data in cfg.successors(offset):
+                block_entry_vars[succ] |= avail
+
+        old_point = new_point
+        new_point = fix_point_progress()
+
+    return block_entry_vars
+
+
+#
+# Analysis related to controlflow
+#
+
+def compute_cfg_from_blocks(blocks):
+    cfg = CFGraph()
+    for k in blocks:
+        cfg.add_node(k)
+
+    for k, b in blocks.items():
+        term = b.terminator
+        for target in _get_targets(term):
+            cfg.add_edge(k, target)
+
+    cfg.set_entry_point(min(blocks))
+    cfg.process()
+    return cfg
+
+
+@singledispatch
+def _get_targets(term):
```

Comment:
addressed in https://github.com/numba/numba/pull/1996/commits/4b0a2f45af5835ac3a6378a0a4a4f577328cba54